## Here we will scrap the codeforces site and check various features of its page and find out the percentage of unrated users.

In [1]:
import requests
page = requests.get("https://codeforces.com/problemset")

from bs4 import BeautifulSoup
soup = BeautifulSoup(page.content, 'html.parser')

In [2]:
print(soup.prettify())

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
 <head>
  <meta content="text/html; charset=utf-8" http-equiv="content-type"/>
  <meta content="49f4e20a7c5dc8e9dbb28ec6863ac600" name="X-Csrf-Token">
   <meta content="width=device-width, initial-scale=0.01" id="viewport" name="viewport"/>
   <script src="//sta.codeforces.com/s/30137/js/jquery-1.8.3.js" type="text/javascript">
   </script>
   <script type="application/javascript">
    window.standaloneContest = false;
        function adjustViewport() {
            var screenWidthPx = Math.min($(window).width(), window.screen.width);
            var siteWidthPx = 1100; // min width of site
            var ratio = Math.min(screenWidthPx / siteWidthPx, 1.0);
            var viewport = "width=device-width, initial-scale=" + ratio;
            $('#viewport').attr('content', viewport);
            var style = $('<style>html * { max-height: 1000000px; }</style>');
            $('html > head').append(style);
        }

        if ( /A

In [5]:
divs = soup.find_all('div' ,  {'class':'content-with-sidebar'})

In [9]:
divs[0]

<div class="content-with-sidebar" id="pageContent">
<div class="second-level-menu">
<ul class="second-level-menu-list">
<li class="current selectedLava"><a href="/problemset">Main</a></li>
<li><a href="/problemsets/acmsguru">acmsguru</a></li>
</ul>
<span style="position:relative;top:-15px;">|</span>
<ul class="second-level-menu-list">
<li><a href="/problemset">Problems</a></li>
<li><a href="/problemset/submit">Submit</a></li>
<li><a href="/problemset/status">Status</a></li>
<li><a href="/problemset/standings">Standings</a></li>
<li><a href="/problemset/customtest">Custom test</a></li>
</ul>
</div>
<style>
    .ProblemRating {
        font-size: 1.1rem;
        /*border: 2px solid #e1e1e1;*/
        /*border-radius: 4px;*/
        padding: 0.2em 0.35em;
        opacity: 0.9;
        font-weight: bold;
    }

    .ProblemRating500, .ProblemRating600, .ProblemRating700, .ProblemRating800, .ProblemRating900,
    .ProblemRating1000, .ProblemRating1100 {
        border-color: #CCC;
    }

  

In [6]:
import re
names = divs[0].find_all('a' , href = re.compile('/problemset/problem/'))

In [18]:
names[3].text

'\r\nPlaylist for Polycarp (easy version)                    '

In [20]:
len(names)

200

In [49]:
divs_new = divs[0].find_all('div' , {'style':'float: left;'} )

In [55]:
names_new = divs_new[0].find_all('a' , href = re.compile('/problemset/problem/'))
names_new[0].text
len(names_new)

1

In [22]:
solved = divs[0].find_all('a' , title = re.compile('Participants solved the problem'))

In [23]:
len(solved)

100

In [37]:
ratings = divs[0].find_all('span' , title = re.compile('Difficulty'))

In [38]:
len(ratings)

91

In [21]:
for i in range(1,200,2):
    print(names[i].text)


Playlist for Polycarp (hard version)                    

Playlist for Polycarp (easy version)                    

Two Pizzas                    

Polycarp and Snakes                    

Extra Element                    

Exam in BerSU (hard version)                    

Exam in BerSU (easy version)                    

Email from Polycarp                    

Ropewalkers                    

Subsequences (hard version)                    

Candy Box (hard version)                    

Topforces Strikes Back                    

Subsequences (easy version)                    

Candy Box (easy version)                    

Computer Game                    

Equalize Prices                    

Nearest Interesting Number                    

Maximum Sine                    

Product Oriented Recurrence                    

Complete Mirror                    

Beautiful Lyrics                    

Plus from Picture                    

Filling Shapes                    

A Story of One

In [34]:
for i in range(0,100):
    num = (solved[i].text)
    print(num[2:])

246
1021
748
1273
3057
3459
6010
6363
8127
646
694
410
1071
4483
4222
6900
8060
109
1043
367
2768
5458
8144
87
177
818
1550
3258
6444
3968
7302
32
200
638
1938
4056
707
1449
600
2688
2005
5511
7443
8870
47
327
878
2239
3285
4414
8101
553
445
2365
5930
6457
7935
5309
7474
27
35
165
274
468
1745
2687
12
18
36
67
161
271
517
657
924
4342
6843
49
100
955
1647
2887
75
626
1104
3950
4916
4597
8974
279
1488
1193
4571
6674
8070
793
742
2922
4478
5656


In [40]:
for i in  range(0,91):
    print(ratings[i].text)

2800
2100
2200
2000
1700
1700
1200
1200
900
2200
2100
2400
2000
1400
1400
1000
900
2900
2200
2500
1700
1300
1000
3000
2600
2100
1900
1500
1000
1500
800
3200
2600
2400
1800
1500
1800
1100
2500
1700
1900
1300
1100
800
2700
2600
2200
1900
1600
1600
1000
2300
2500
1900
1300
1200
1000
1100
700
3300
3300
2800
2700
2500
2000
1800
1500
1000
3100
2900
2400
2000
1800
3000
2300
2100
1500
1400
1400
700
2800
2000
2100
1600
1100
900
2400
2400
1800
1600
1300


In [42]:
pages = soup.find_all('span' , {'class':'page-index'})
len(pages)

5

In [43]:
pages[4]

<span class="page-index" pageindex="52"><a href="/problemset/page/52">52</a></span>

In [2]:
import requests
page = requests.get("https://codeforces.com/problemset/page/52")

from bs4 import BeautifulSoup
soup = BeautifulSoup(page.content, 'html.parser')

In [45]:
pages = soup.find_all('span' , {'class':'page-index'})
len(pages)

5

In [46]:
pages[4]

<span class="page-index active" pageindex="52"><a href="/problemset/page/52">52</a></span>

In [7]:
lists = []
for i in range(1,53):
    url = 'https://codeforces.com/problemset/page/' + str(i)
    lists.append(url)
    print(url)

https://codeforces.com/problemset/page/1
https://codeforces.com/problemset/page/2
https://codeforces.com/problemset/page/3
https://codeforces.com/problemset/page/4
https://codeforces.com/problemset/page/5
https://codeforces.com/problemset/page/6
https://codeforces.com/problemset/page/7
https://codeforces.com/problemset/page/8
https://codeforces.com/problemset/page/9
https://codeforces.com/problemset/page/10
https://codeforces.com/problemset/page/11
https://codeforces.com/problemset/page/12
https://codeforces.com/problemset/page/13
https://codeforces.com/problemset/page/14
https://codeforces.com/problemset/page/15
https://codeforces.com/problemset/page/16
https://codeforces.com/problemset/page/17
https://codeforces.com/problemset/page/18
https://codeforces.com/problemset/page/19
https://codeforces.com/problemset/page/20
https://codeforces.com/problemset/page/21
https://codeforces.com/problemset/page/22
https://codeforces.com/problemset/page/23
https://codeforces.com/problemset/page/24
h

In [57]:
divs = soup.find_all('div' ,  {'class':'content-with-sidebar'})
divs_new = divs[0].find_all('div' , {'style':'float: left;'} )
total = 0
for i in range(0,52):
    page = requests.get(lists[i])
    soup = BeautifulSoup(page.content, 'html.parser')
    divs_new = divs[0].find_all('div' , {'style':'float: left;'} )
    l = len(divs_new)
    total = total + l
    
print(total)    

4212


In [5]:
page = requests.get('https://codeforces.com/problemset/standings/page/1609')
soup = BeautifulSoup(page.content, 'html.parser')

In [60]:
divs = soup.find_all('a' , title = re.compile('Unrated') )
len(divs)

152

In [8]:
lists = []
for i in range(1,1610):
    url = 'https://codeforces.com/problemset/standings/page/' + str(i)
    lists.append(url)
    print(url)

https://codeforces.com/problemset/standings/page/1
https://codeforces.com/problemset/standings/page/2
https://codeforces.com/problemset/standings/page/3
https://codeforces.com/problemset/standings/page/4
https://codeforces.com/problemset/standings/page/5
https://codeforces.com/problemset/standings/page/6
https://codeforces.com/problemset/standings/page/7
https://codeforces.com/problemset/standings/page/8
https://codeforces.com/problemset/standings/page/9
https://codeforces.com/problemset/standings/page/10
https://codeforces.com/problemset/standings/page/11
https://codeforces.com/problemset/standings/page/12
https://codeforces.com/problemset/standings/page/13
https://codeforces.com/problemset/standings/page/14
https://codeforces.com/problemset/standings/page/15
https://codeforces.com/problemset/standings/page/16
https://codeforces.com/problemset/standings/page/17
https://codeforces.com/problemset/standings/page/18
https://codeforces.com/problemset/standings/page/19
https://codeforces.co

In [13]:
tot_a = 0
tot_r = 0
tot_u = 0
for i in range(0,1609):
    page = requests.get(lists[i])
    soup = BeautifulSoup(page.content, 'html.parser')
#     all_members = 200#soup.find_all('a' , href = re.compile('/profile/') )
    unrated = soup.find_all('a' , title = re.compile('Unrated') )
#     a = 200 #len(all_members)
    u = len(unrated)
#     rated = a - u
#     tot_a = tot_a + a
    tot_u = tot_u + u
#     tot_r = tot_r + rated
    
# print(tot_a)  
print(tot_u/321889*100)

43.980378329175586


### Hence 43.980378329175586% of users in codeforces are unrated where 321889 are the total number of users.